File: perl/devel_misc/ref.java
class A {

  public A() {
    System.out.println("void constructor called");
  }

  public A(int v) {
    this.v = v;
  }

  public A(A a) {
    v = a.v+1;
  }

  public void inc() {
    v++;
  }

  public void clone(A a) {
    v = a.v;
  }

  public void print() {
    System.out.println(v);
  }

  int v;
}

public class ref {
  static void swap(A a1, A a2) {
    A a3 = new A(1000);
    
    a3.clone(a1);
    a1.clone(a2);
    a2.clone(a3);
  }

  public static void main(String[] args) {
    A a = new A(100);
    A b = new A(10);

    A c = a;

    a.inc();

    a.print();
    b.print();
    c.print();

    A a1 = new A(15);
    A a2 = new A(35);

    a1.print();
    a2.print();

    swap(a1, a2);

    a1.print();
    a2.print();

  }
}

File: perl/devel_misc/ref.cpp
#include <iostream>
using namespace std;

class A {
  static int ord;
public:
  A() {
    v = ++ord;
    cout << "void constructor called with " << v << "\n";
  }

  A(const A& a) {
    cout << "copy constructor called with " << a.v << "\n";
    v = a.v;
  }

  A& operator=(const A& a) {
    cout << "assignment operator called with " << a.v << "\n";
    v = a.v;
    return *this;
  }

  int v;
};

int A::ord = 0;

void f(A& a) {
  A* ap = new A();
  cout << " now assigning...\n";
  a = *ap; 
   cout << " assigning done...\n";

}

void fp(A *ap) {
  A *ap2 = new A();
  cout << "now assigning...\n";
  *ap = *ap2;
  cout << "assigning done.\n";
}

void swap(A& a1, A& a2) {
  A t = a1;

  a1 = a2;
  a2 = t;
}


int main()
{
  A* ap = new A();
  A a = *ap;

  cout << a.v << "\n";

  f(a);

  cout << a.v << "\n";

  fp(&a);

  cout << a.v << "\n";

  cout << "swap test\n";

  A a1, a2;

  swap(a1, a2);


}

